### PR TITLE
Ensure checksum type is always present

### PR DESCRIPTION
**Pull request by noirbizarre**
This is the first part of fix on checksum handling (and globally on resources saving and validations).
This one fix bad checksum in database and raise ValidationError if not present ensuring it's not happening again.

### CODE REVIEW

**Comment by davidbgk** [comment]
r+

**Comment by davidbgk** [comment]
Note that tests have failed.

File: udata/tests/__init__.py [hunk 1]
```diff
@@ -176,6 +176,7 @@ def init_search(self):
         self._used_search = True
         self.app.config['AUTO_INDEX'] = True
         es.initialize()
+        es.cluster.health(wait_for_status='yellow', request_timeout=3)
```
Comment (davidbgk) [inline udata/tests/__init__.py]: :+1:

File: udata/tests/api/test_datasets_api.py [hunk 2]
```diff
@@ -93,6 +96,7 @@ def test_dataset_api_get_deleted_but_authorized(self):
         response = self.get(url_for('api.dataset', dataset=dataset))
         self.assert200(response)
 
+    @attr('create')
```
Comment (davidbgk) [inline udata/tests/api/test_datasets_api.py]: I didn't know about that feature, nice!

File: udata/core/dataset/forms.py [hunk 3]
```diff
@@ -22,27 +22,6 @@ class ChecksumForm(ModelForm):
     value = fields.StringField()
 
 
-class ChecksumWidget(object):
```
Comment (davidbgk) [inline udata/core/dataset/forms.py]: We don't need a field/widget anymore?

File: udata/forms/fields.py [hunk 4]
```diff
@@ -200,10 +198,29 @@ def __getattr__(self, name):
...
                                         *args, **kwargs)
 
+    def process(self, formdata, data=unset_value):
+        self._formdata = formdata
+        return super(FormField, self).process(formdata, data=data)
+
+    def validate(self, form, extra_validators=tuple()):
+        if extra_validators:
+            raise TypeError('FormField does not accept in-line validators, '
+            'as it gets errors from the enclosed form.')
```
Comment (davidbgk) [inline udata/forms/fields.py]: Better indent at the opened parenthesis?

File: udata/forms/fields.py [hunk 5]
```diff
@@ -200,10 +198,29 @@ def __getattr__(self, name):
...
     def __init__(self, form_class, *args, **kwargs):
+        self.has_data = False
         super(FormField, self).__init__(FormWrapper(form_class),
                                         *args, **kwargs)
 
+    def process(self, formdata, data=unset_value):
+        self._formdata = formdata
+        return super(FormField, self).process(formdata, data=data)
+
+    def validate(self, form, extra_validators=tuple()):
```
Comment (davidbgk) [inline udata/forms/fields.py]: Initialize with `None` and check `if extra_validators is None`.

File: udata/forms/fields.py [hunk 6]
```diff
@@ -200,10 +198,29 @@ def __getattr__(self, name):
 
 class FormField(FieldHelper, fields.FormField):
     def __init__(self, form_class, *args, **kwargs):
+        self.has_data = False
```
Comment (davidbgk) [inline udata/forms/fields.py]: I'm not familiar enough with WTForm but that looks weird that you have to deal with that manually :/

File: udata/forms/fields.py [hunk 7]
```diff
@@ -200,10 +198,29 @@ def __getattr__(self, name):
...
+        self._formdata = formdata
+        return super(FormField, self).process(formdata, data=data)
+
+    def validate(self, form, extra_validators=tuple()):
+        if extra_validators:
+            raise TypeError('FormField does not accept in-line validators, '
+            'as it gets errors from the enclosed form.')
+
+        # Run normal validation only if there is data for this form
+        prefix = '{0}-'.format(self.name)
```
Comment (davidbgk) [inline udata/forms/fields.py]: Move that to the `init` for consistency with `NestedModelList`?

File: udata/forms/fields.py [hunk 8]
```diff
@@ -391,16 +408,29 @@ def __init__(self, model_form, *args, **kwargs):
...
-        if formdata and any(k.startswith(prefix) for k in formdata):
+        self.has_data = formdata and any(k.startswith(self.prefix) for k in formdata)
+        if self.has_data:
             super(NestedModelList, self).process(formdata, data)
         else:
-            super(NestedModelList, self).process(None, data)
+            self.entries = []
+            # super(NestedModelList, self).process(None, data)
+
+    def validate(self, form, extra_validators=tuple()):
```
Comment (davidbgk) [inline udata/forms/fields.py]: Same here.

File: udata/forms/fields.py [hunk 9]
```diff
@@ -391,16 +408,29 @@ def __init__(self, model_form, *args, **kwargs):
...
         self.data_submitted = False
         self.initial_data = []
+        self.prefix = '{0}-'.format(self.name)
 
     def process(self, formdata, data=unset_value):
+        self._formdata = formdata
         self.initial_data = data
-        prefix = '{0}-'.format(self.name)
-        if formdata and any(k.startswith(prefix) for k in formdata):
+        self.has_data = formdata and any(k.startswith(self.prefix) for k in formdata)
```
Comment (davidbgk) [inline udata/forms/fields.py]: It smells logic duplication.

**Comment by davidbgk** [comment]
So great to have hunt that bug down, r+wc :)

File: udata/tests/__init__.py (hunk 1 above)
Comment (noirbizarre) [inline udata/tests/__init__.py]: It was driving me crazy! Such time lost !!!

File: udata/tests/api/test_datasets_api.py (hunk 2 above)
Comment (noirbizarre) [inline udata/tests/api/test_datasets_api.py]: I just discovered it ! Very useful !

File: udata/core/dataset/forms.py (hunk 3 above)
Comment (noirbizarre) [inline udata/core/dataset/forms.py]: We don't have front forms any more so all widgets are no longer used (except for the coming login/register)

File: udata/forms/fields.py (hunk 4 above)
Comment (noirbizarre) [inline udata/forms/fields.py]: Sorry, dumb copy past from upstream, I will fix that.

File: udata/forms/fields.py (hunk 5 above)
Comment (noirbizarre) [inline udata/forms/fields.py]: I kept the upstream signature. I think there is some iteration in `super().validate()`

File: udata/forms/fields.py (hunk 6 above)
Comment (noirbizarre) [inline udata/forms/fields.py]: The big problem is:
- WTForms is meant for form validation (so flat structure with MultiDict) and we are dealing with JSON and nested object
- WTForms consider FormField as required. It took me a lot of mailinglists/issues/stackoverflow reading to come to this "clean" solution :/

Conclusion: I think we should migrate from WTForms to something dedicated to JSON/dict

File: udata/forms/fields.py (hunk 7 above)
Comment (noirbizarre) [inline udata/forms/fields.py]: OK

File: udata/forms/fields.py (hunk 9 above)
Comment (noirbizarre) [inline udata/forms/fields.py]: Yep, but I didn't want to factorize yet. It seems overkill to me for 2 attributes.
It's to becoming complex :/

**Pull request merged**